### ISSUE TITLE
"reportUnusedExpression" Warning Not Triggered for Unused Methods

### ISSUE THREAD

**Issue by ilotoki0804** [microsoft/pyright#5958]
**Bug Description**

The Pylance warning "reportUnusedExpression" is designed to trigger when an expression value is unused. However, this warning does not seem to be triggered in the context of class and module methods.

**Code Examples and Screenshots**
```python
# Built-in object

my_list = [4, 2, 8, 3]

my_list.sort  # A warning should be triggered here (but it's not)

my_list  # The warning is correctly triggered here (warning: Expression value is unused Pylance (reportUnusedExpression))

# User-defined object

class Hello:
    def bar(self, a) -> str:
        return 'baz'

hello = Hello()

hello.bar  # A warning should be triggered here (but it's not)

Hello.bar  # A warning should be triggered here (but it's not)

# Module

import requests

requests.get  # A warning should be triggered here (but it's not)

requests.get('https://example.com')  # No error occurred (possibly intended?)

# Normal environment

1234  # The warning is correctly triggered here (warning: Expression value is unused Pylance (reportUnusedExpression))

def foo() -> int:
    return 0

foo   # The warning is correctly triggered here(warning: Expression value is unused Pylance(reportUnusedExpression))
```

requests: 2.31.0

**VS Code Extension or Command-Line Usage**

Pylance version: v2023.9.10

**Comment by erictraut** [comment]
`reportUnusedExpression` is working as designed here, so I don't consider this a bug. I'm going to switch this to an enhancement request and consider it in that light.

The `reportUnusedExpression` diagnostic check was added in response to [this feature request](https://github.com/microsoft/pyright/issues/3174), which also includes the thinking behind its design and limitations.

It was later extended by [this enhancement request](https://github.com/microsoft/pyright/issues/3896).

When considering an enhancement request, we need to consider the upsides and downsides of the request. The `reportUnusedExpression` is on by default even in "basic" type checking mode, so we need to be conservative about producing false positive errors. There is some concern that the enhancement you're suggesting will result in false positive errors. Member access expressions in Python (e.g. `hello.bar` or `requests.get`) do not typically have side effects, but they can in some cases. In those cases, it would be legitimate to use such an expression without assigning its value.

I'll need to give this more thought and perhaps do some experimentation.

**Comment by erictraut** [comment]
I've prototyped the suggested change and ran this on a bunch of existing typed code bases. As I feared, this change produces a lot of [false positive errors](https://github.com/microsoft/pyright/pull/5961), especially in test code.

**Comment by ilotoki0804** [comment]
> I've prototyped the suggested change and ran this on a bunch of existing typed code bases. As I feared, this change produces a lot of [false positive errors](https://github.com/microsoft/pyright/pull/5961), especially in test code.

In fact, it's fairly common that Pylance produces errors when `try-except`, or other error handling mechanisms, is used in the code.
How about encourage them to use `hasattr` or `getattr`?

```python
# Source: https://github.com/spack/spack/blob/develop/lib/spack/spack/spec.py#L2978

# Now
try:
    v.ref_version
except vn.VersionLookupError:
    ...

# Using hasattr
if hasattr(v, 'ref_version'):
    ...
```
```python
# Source: https://github.com/spack/spack/blob/develop/lib/spack/spack/modules/common.py#L877

# Now
try:
    self.default_template
except AttributeError:
    ...

# Using hasattr
if not hasattr(self, 'default_template'):
   ...
```
```python
# Source: https://github.com/FasterSpeeding/Tanjun/blob/master/tests/test_checks.py#L2470

# now
with pytest.raises(AttributeError):
    command.wrapped_command

# Using hasattr
assert not hasattr(command, 'wrapped_command')
```
Not only does it resolve the issue, but it also looks more concise for me.
I haven't reviewed all the failed code, so please inform me if there's any that can't be addressed using `hasattr` or `getattr`.

**Comment by erictraut** [comment]
If this check were enabled only in strict mode, there would be more latitude for encouraging code changes. Since this check is enabled in basic mode, it affects users who are not fully bought in to the benefits of static type checking and are sensitive to false positives and the need to change their code to satisfy a type checker.

The other thing to note is that `reportUnusedExpression` isn't reporting a real bug. Most of the time it's telling you about "messy lint that you might wan to clean up" and "something that looks suspicious that you might want to know about".
